File: frontend/src/lib/offline/health.ts
import { API_URL } from '../api/client'

// Sonda de saúde (offline-pwa 4.3 / D7-4). `HEAD /api/v1/health` é o PORTEIRO da
// drenagem: um toque barato antes de disparar os envios, para que um Wi-Fi de
// galpão sem rota de saída (associado mas sem internet) produza UMA sonda e não 40
// requisições. Qualquer falha de rede ou status não-ok = servidor inalcançável.

export async function probeHealth(
  deps: { fetchImpl?: typeof fetch; baseUrl?: string; signal?: AbortSignal } = {},
): Promise<boolean> {
  const fetchImpl = deps.fetchImpl ?? (typeof fetch !== 'undefined' ? fetch : undefined)
  if (!fetchImpl) return false
  const baseUrl = deps.baseUrl ?? API_URL
  try {
    const res = await fetchImpl(`${baseUrl}/api/v1/health`, { method: 'HEAD', signal: deps.signal })
    return res.ok
  } catch {
    return false
  }
}

File: backend/app/controllers/api/v1/health.rb
# frozen_string_literal: true

module Api
  module V1
    # Sonda de saúde (offline-pwa 4.3 / D7-4). A fila offline usa `HEAD
    # /api/v1/health` como PORTEIRO da drenagem: um único toque barato antes de
    # disparar os envios, para que um Wi-Fi de galpão sem rota de saída produza
    # UMA sonda e não 40 requisições. Sem tenant e sem auth — é só "o servidor
    # responde?". Grape atende HEAD por meio do handler de GET (corpo descartado).
    class Health < Grape::API
      format :json

      resource :health do
        desc 'Sonda de saúde — 200 quando o servidor está de pé'
        get do
          status 200
          { status: 'ok' }
        end
      end
    end
  end
end
